client/site-profiler/components/get-report-form/index.tsx

```diff
@@ -0,0 +1,233 @@
+import emailValidator from 'email-validator';
+import { translate } from 'i18n-calypso';
+import { FormEvent, useState } from 'react';
+import FormFieldset from 'calypso/components/forms/form-fieldset';
+import FormTextInput from 'calypso/components/forms/form-text-input';
+import wp from 'calypso/lib/wp';
+import './styles.scss';
+
+type Errors = {
+	name?: string;
+	email?: string;
+	termsAccepted?: string;
+};
+
+export function GetReportForm( {
+	url,
+	token,
+	onClose,
+}: {
+	url?: string;
+	token?: string;
+	onClose: () => void;
+} ) {
+	const [ name, setName ] = useState( '' );
+	const [ email, setEmail ] = useState( '' );
+	const [ isTermsChecked, setIsTermsChecked ] = useState( false );
+	const [ errors, setErrors ] = useState< Errors | null >( null );
+	const [ responseError, setResponseError ] = useState< string | null >( null );
+	const [ isSubmitting, setIsSubmitting ] = useState( false );
+	const [ responseSuccess, setResponseSuccess ] = useState( false );
+
+	const handleSubmit = async ( e: FormEvent< HTMLFormElement > ) => {
+		e.preventDefault();
+
+		const formData = new FormData( e.currentTarget );
+		const errors = validateForm( formData );
+		if ( errors ) {
+			setErrors( errors );
+			return;
+		}
+		setErrors( null );
+		setIsSubmitting( true );
+
+		let response = null;
+		try {
+			response = await wp.req.post(
+				{
+					path: '/site-profiler/lead',
+					apiNamespace: 'wpcom/v2',
+				},
+				{
+					name: formData.get( 'name' ),
+					email: formData.get( 'email' ),
+					url,
+					hash: token,
+				}
+			);
+		} catch ( error ) {
+			setResponseError( error instanceof Error ? error.message : String( error ) );
+			return;
+		} finally {
+			setIsSubmitting( false );
+		}
+		if ( response.success ) {
+			setResponseSuccess( true );
+		}
+	};
+
+	const validateName = ( name: string ) => {
+		if ( ! name ) {
+			return translate( 'Name is required' );
+		}
+		return null;
+	};
+
+	const validateEmail = ( email: string, fullValidation = false ) => {
+		if ( ! email ) {
+			return translate( 'Email is required' );
+		}
+		if ( fullValidation && ! emailValidator.validate( email ) ) {
+			return translate( 'Please enter a valid email address' );
+		}
+		return null;
+	};
+
+	const validateTerms = ( terms: boolean ) => {
+		if ( ! terms ) {
+			return translate( 'Terms must be accepted' );
+		}
+		return null;
+	};
+
+	function validateForm( formData: FormData ) {
+		let errors = {};
+		let hasErrors = false;
+
+		let error = validateName( formData.get( 'name' ) as string );
+		if ( error ) {
+			errors = { ...errors, name: error };
+			hasErrors = true;
+		}
+		error = validateEmail( formData.get( 'email' ) as string, true );
+		if ( error ) {
+			errors = { ...errors, email: error };
+			hasErrors = true;
+		}
+		error = validateTerms( Boolean( formData.get( 'termsAccepted' ) ) );
+		if ( error ) {
+			errors = { ...errors, termsAccepted: error };
+			hasErrors = true;
+		}
+		return hasErrors ? errors : null;
+	}
```

Comment:
Considering this and the others validation functions are pure functions, I would put them outside the component, and preferably on other files.

I think this would make this component easier to read and the functions easier to test. What do you think?

Reply:
I totally agree, I started adding them to the file for simplicity, but they have reached a size that makes it better to extract to an external file in the same folder, as it is very related to this form. I also considered extracting common functionality to "agnostic" functions independent of the field, but I don't think is worthy for a form with three fields 😄